### DISCUSSION TITLE
[Compilers] <Troubles with setting up zksolc in Forge to the newest version>

### DISCUSSION THREAD

**Discussion by Szczepoo13** [zkSync-Community-Hub/zksync-developers#737] [Compilers] [answered]
### Which compiler are you using?

- [ ] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.20

### zksolc/zkvyper version?

1.5.4

### Explain the issue

I am trying to fork, compile, and then deploy the **https://github.com/aave-dao/aave-v3-origin** instance locally on my zkSync chain built within zkStack. But while compiling a contracts I have many errors/challenges one of them has been mentioned here ( https://github.com/matter-labs/hardhat-zksync/issues/1475), 
However also I have got (without --zksync flag everything works): 
![image](https://github.com/user-attachments/assets/f066fcf5-26a5-4019-adb0-17032d7391c8)

Since it seems like Foundry-zksync has the newest version of compiler up to 1.5.4 I also tried to update to the new version of zksolc 1.5.6 with the build binaries from sources by following the tutorial: https://github.com/matter-labs/era-compiler-solidity, however after pointing to new 1.5.6 compiler with (**forge build --zk-solc-path ./target/*/zksolc** --via-ir --zksync) I was getting the error "Stack too deep". 
![stackdeep](https://github.com/user-attachments/assets/4bd1edda-0dcf-401f-b6c2-d00d3a3e2cbc).

Apart from the errors mentioned above, my question is: how do I point Forge to use the currently newest version of zksolc (1.5.6), instead of 1.5.4??. Doing so I think would solve most of the issues above. 

I have also tried the other way around with downgrading the zksolc version to for example version 1.5.3: **foundryup-zksync --version nightly-a1155ab78056eca730343d4268b851750e453166**   but I am getting the following errors.
![downgrade1](https://github.com/user-attachments/assets/8188e4b0-061c-4420-8fe2-14af796844cf)

Note: The newest instance of aave-v3-origin is Foundry-based (project structure), switching for Hardhat would be quite challenging.

**Comment by alhenaai** [comment]
To point Forge to use the newest version of zksolc (1.5.6) instead of 1.5.4, you can specify the path to the new compiler binary directly in your command. You mentioned using the command:

```
forge build --zk-solc-path ./target/*/zksolc --via-ir --zksync
```

This is generally the correct approach. Ensure that the path you provide (`./target/*/zksolc`) accurately points to the `zksolc` binary of version 1.5.6 that you compiled from sources. Here are a few steps to troubleshoot and ensure it works correctly:

1. **Verify the Path**: Make sure that the path to `zksolc` is correct and that the binary is executable. You can check this by running:
   ```bash
   ls -l ./target/*/zksolc
   ```
   This command should list the `zksolc` binary if the path is correct.

2. **Permissions**: Ensure that the `zksolc` binary has the necessary execution permissions:
   ```bash
   chmod +x ./target/*/zksolc
   ```

3. **Direct Invocation**: Try invoking `zksolc` directly from the command line to see if it runs without errors:
   ```bash
   ./target/*/zksolc --version
   ```
   This should return the version of the compiler.

4. **Forge Command**: If the above steps are correct, your Forge command should ideally work. If it still doesn't, consider clearing any caches or temporary files that might be causing issues.

5. **Check for Updates**: Sometimes, tooling issues might be due to bugs that are fixed in newer releases. Regularly check the [Foundry-zksync GitHub repository](https://github.com/matter-labs/foundry-zksync) for updates or patches that might address your issue.

If you continue to experience issues even after ensuring the correct compiler path and version, it might be helpful to open an issue on the Foundry-zksync GitHub page or seek further assistance from the community or documentation specific to these tools.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by Szczepoo13** [reply]
After typing zksolc --version, it shows: Solidity compiler for ZKsync v1.5.6 (LLVM build 801623bd56fbf4e1faf964d4c8dbce563bed9384). However while running command **forge build --zk-solc-path ./target/*/zksolc --via-ir --zksync** it gives an error: version not found in Solc output: ![VersionNotFound](https://github.com/user-attachments/assets/1eae8467-9ce7-4a11-b66a-5d95c90c9543). What could be missing/wrong?

**Reply by alhenaai** [reply]
The error "version not found in Solc output" suggests that there might be an issue with how the `forge` tool is interpreting the output from the `zksolc` compiler. Here are a few steps to troubleshoot and potentially resolve this issue:

1. **Check Compiler Integration**: Ensure that `forge` is compatible with the custom `zksolc` version you are using. Sometimes, the tooling around custom compilers may expect specific output formats or flags.

2. **Compiler Output Verification**: Run `zksolc` with a simple Solidity file to check the output format. You can create a simple Solidity file named `Test.sol` with the following content:
   ```solidity
   // SPDX-License-Identifier: MIT
   pragma solidity ^0.8.0;

   contract Test {
       function test() public pure returns (string memory) {
           return "Hello, World!";
       }
   }
   ```
   Then run:
   ```bash
   ./target/*/zksolc Test.sol
   ```
   Check if the output includes the version information in a format that `forge` might expect.

3. **Forge Configuration**: Review the `foundry.toml` or any relevant configuration files to ensure there are no misconfigurations or incorrect paths that might be causing `forge` to fail in reading the compiler version.

4. **Verbose Logging**: Run the `forge` command with verbose logging to get more detailed error information. This can often provide insights into what the tool is actually doing:
   ```bash
   forge build --zk-solc-path ./target/*/zksolc --via-ir --zksync --verbose
   ```

5. **Consult Documentation**: Check the [Foundry documentation](https://book.getfoundry.sh/) and the [zksolc GitHub repository](https://github.com/matter-labs/zksolc-bin) for any notes on compatibility or special configuration requirements when using custom compilers.

6. **Community and Support**: If the issue persists, consider reaching out on the [Foundry Discord](https://discord.gg/foundry) or filing an issue on the [GitHub repository](https://github.com/matter-labs/foundry-zksync/issues) for more targeted support. Sometimes, issues like these can stem from bugs or unhandled cases in the tools themselves.

By following these steps, you should be able to identify and potentially resolve the issue with the version not being found in the Solc output.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by uF4No** [comment]
Have you tried adding it to the `foundry.toml`?

```
[profile.default.zksync]
zksolc = '1.5.6'
```

**Reply by Szczepoo13** [reply]
Thanks for your reply! Yes, I have tried it already, but the outcome is the same 💯 
![foundrytoml](https://github.com/user-attachments/assets/275dfcdc-3856-4d87-adf2-49d699e499ef)

**Reply by uF4No** [reply, marked as answer]
must be inside `profile.default.zksync` though 👀. And if it's in the `foundry.toml` there's no need to specify it with `--zk-solc-path`.

**Reply by uF4No** [reply]
Btw, given that you're looking to deploy Aave yourself, you can check out the repos that the BGD team created for the ZKsync Era deployment:

https://github.com/bgd-labs/aave-delivery-infrastructure
https://github.com/bgd-labs/aave-v3-zksync-proposal-simulation
https://github.com/bgd-labs/aave-helpers
https://github.com/bgd-labs/aave-address-book
https://github.com/bgd-labs/aave-proposals-v3

**Reply by Szczepoo13** [reply]
Oh, man! I marked it as the answer. All that troubleshooting just to edit one line with .zksync, and it worked! Thank you so much for spotting that :))

**Reply by Karrq** [reply]
Hey @Szczepoo13 you should note that `--zk-solc-path` is _not_ the path for `zksolc`, but rather the path to the `solc` compiler that `zksolc` should use.
As @uF4No pointed out, the config option `zksync.zksolc` is where you want to specify which `zksolc` to use, as there's no CLI flag for it.
The equivalent configuration option for `--zk-solc-path` is `zksync.solc_path`, which is a bit more explicit on what is being configured I believe
